feat:Refactored serach input component

albinAppsmith opened the pull request:
## Description

Added refactored search input.
Also this PR includes renaming of as to renderAs. This is needed because of the conflict with styled component as prop.

Fixes # (issue)
https://github.com/appsmithorg/appsmith/issues/21250

Depends on # (pr)

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

- Manual on storybook 
- Manual on main repo
- Jest
- Cypress

### Test Plan
> Add Testsmith test cases links that relate to this PR

### Issues raised during DP testing
> Link issues raised during DP testing for better visiblity and tracking (copy link from comments dropped on this PR)


## Checklist:
### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes


### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

tanvibhakta commented on packages/design-system/src/Input/Input.styles.tsx:
```diff
@@ -38,7 +38,7 @@ const getSizes = (size: InputSizes, component: "input" | "textarea") => {
   labelPosition?: "top" | "left";
-  size?: InputSizes;
+  testSize?: InputSizes;
   component: "input" | "textarea";
```
What's a testSize?

tanvibhakta commented on packages/design-system/src/Input/Input.tsx:
```diff
@@ -24,67 +28,98 @@ import {
+      className,
       description,
       endIcon,
+      endIconProps,
```
What is the purpose of doing this?

albinAppsmith replied:
For passing down the `onClick` function.

albinAppsmith replied:
Will refactor this later and implement icon button here

shadabbuchh commented:
### Bugs in this PR:

**Functional:**

1. className & value do not render at all.
2. Should the desc of className be (try not to) pass additional classes here instead of what it is right now?
3. Can't see the active state alone. Active & focus state render at the same time.
4. Should we introduce a control that changes the placeholder?
5. We see the x icon in the default state also even when no text is entered.

**Design:**

1. Length is 250px (expected is 206px). Is this auto
2. Wrong colour of icon in Default state
3. Wrong top & bottom padding of the magnifying glass icon
4. Wrong gap between text/placeholder & magnifying glass
5. No box shadow in any state
6. Wrong background colour in Hover state
7. Wrong background colour in entered -> no focus state
8. Wrong top & bottom padding for x icon.

albinAppsmith commented:
1. className & value do not render at all => fixed className issue. For value you need to reload page. This is something storybook has to handle.
2. Should the desc of className be (try not to) pass additional classes here instead of what it is right now? => Changed
3. Can't see the active state alone. Active & focus state render at the same time => This is expected behaviour
4. Should we introduce a control that changes the placeholder? => Good catch, added that
5. We see the x icon in the default state also even when no text is entered. => Good catch, fixed.

1. Length is 250px (expected is 206px). Is this auto => Yes this is auto.
Rest of the behaviours like hover and all will 
2. Wrong colour of icon in Default state => Consider the design of input for this, as discussed with Vinay.
3. Wrong top & bottom padding of the magnifying glass icon => Consider the design of input for this, as discussed with Vinay.
4. Wrong gap between text/placeholder & magnifying glass => Consider the design of input for this, as discussed with Vinay.
5. No box shadow in any state => Not required this because it is not consistent with other inputs.
6. Wrong background colour in Hover state => Consider the design of input for this, as discussed with Vinay.
7. Wrong background colour in entered -> no focus state => Consider the design of input for this, as discussed with Vinay.
8. Wrong top & bottom padding for x icon. => Consider the design of input for this, as discussed with Vinay.


@shadabbuchh Please re verify.

shadabbuchh commented:
All the identified issues have been fixed. This is good to go.
